src/Adapter/Product/Update/ProductSupplierUpdater.php

```diff
@@ -63,42 +69,107 @@ class ProductSupplierUpdater
     }
 
     /**
      * @param ProductId $productId
      * @param SupplierId $defaultSupplierId
-     * @param ProductSupplier[] $productSuppliers
+     * @param array<int, ProductSupplier> $productSuppliers
      *
-     * @throws CannotUpdateProductSupplierException
-     * @throws CoreException
-     * @throws ProductSupplierNotFoundException
+     * @return array<int, ProductSupplierId>
      */
-    public function setProductSuppliers(ProductId $productId, SupplierId $defaultSupplierId, array $productSuppliers): void
-    {
-        $deletableProductSupplierIds = $this->getDeletableProductSupplierIds($productId->getValue(), $productSuppliers);
+    public function setProductSuppliers(
+        ProductId $productId,
+        SupplierId $defaultSupplierId,
+        array $productSuppliers
+    ): array {
+        $product = $this->productRepository->get($productId);
 
-        foreach ($productSuppliers as $productSupplier) {
-            if ($productSupplier->id) {
-                $this->productSupplierRepository->update($productSupplier);
-            } else {
-                $this->productSupplierRepository->add($productSupplier);
-            }
+        if ($product->cache_default_attribute) {
+            throw new InvalidProductTypeException(sprintf(
+                'Product #%d has combinations. Use %s::%s to set product suppliers for specified combination',
+                $productId->getValue(),
+                self::class,
+                'setCombinationSuppliers()'
+            ));
         }
 
-        $this->productSupplierRepository->bulkDelete($deletableProductSupplierIds);
+        $this->persistProductSuppliers($productId, $productSuppliers);
         $this->updateDefaultSupplier($productId, $defaultSupplierId);
+
+        return $this->getCurrentProductSupplierIds($productId);
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param CombinationId $combinationId
+     * @param array<int, ProductSupplier> $productSuppliers
+     *
+     * @return array<int, ProductSupplierId>
+     */
+    public function setCombinationSuppliers(
+        ProductId $productId,
+        CombinationId $combinationId,
+        array $productSuppliers
+    ): array {
+        $this->persistProductSuppliers($productId, $productSuppliers, $combinationId);
+
+        // make sure all non-combination suppliers are deleted
+        $existingNonCombinationSuppliers = $this->getCurrentProductSupplierIds($productId);
+        $this->productSupplierRepository->bulkDelete($existingNonCombinationSuppliers);
+
+        return $this->getCurrentProductSupplierIds($productId, $combinationId);
+    }
+
+    /**
+     * Removes all product suppliers associated to specified product without combinations
+     *
+     * @param ProductId $productId
+     */
+    public function removeAllForProduct(ProductId $productId): void
+    {
+        $product = $this->productRepository->get($productId);
+
+        if ($product->cache_default_attribute) {
```

Comment:
```suggestion
        if ($product->hasCombinations()) {
```

Same suggestion (optional)